### PR TITLE
help: improve writing classes

### PR DESCRIPTION
**Pull request by telephon**
This is a next step in improving the Writing Classes helpfile.

### CODE REVIEW

File: HelpSource/Guides/WritingClasses.schelp [hunk 1]
```diff
@@ -284,9 +285,11 @@ methods should take only one argument to support both ways of expression consist
 
 code::
 MyClass {
-    variable_ { | newValue |
+
+	variable_ { | newValue |
         variable = newValue.clip(minval,maxval);
```
Comment (gusano) [inline HelpSource/Guides/WritingClasses.schelp]: @telephon there is a mix of spaces and tabs in this method declaration and other few ones

File: HelpSource/Guides/WritingClasses.schelp [hunk 2]
```diff
@@ -295,10 +298,12 @@ not return the new value from your setter.
 
 code::
 MyClass {
-    variable_ { | newValue |
+
+	variable_ { | newValue |
```
Comment (gusano) [inline HelpSource/Guides/WritingClasses.schelp]: @telephon here as well

File: HelpSource/Guides/WritingClasses.schelp [hunk 3]
```diff
@@ -364,7 +381,9 @@ code::
 MyTestPoint {
     var <x, <y;
 
-    *new{ |x, y| ^super.newCopyArgs(x, y) }
+    *new { |x, y|
+		^super.newCopyArgs(x, y)
+	}
```
Comment (gusano) [inline HelpSource/Guides/WritingClasses.schelp]: @telephon indention looks wrong :)

File: HelpSource/Guides/WritingClasses.schelp [hunk 4]
```diff
@@ -389,27 +408,33 @@ code::
 MyTestPoint {
     var <x, <y;
 
-    *new{ |x, y| ^super.newCopyArgs(x,y) }
+    *new { |x, y|
+		^super.newCopyArgs(x,y)
+	}
```
Comment (gusano) [inline HelpSource/Guides/WritingClasses.schelp]: @telephon same here

File: HelpSource/Guides/WritingClasses.schelp [hunk 5]
```diff
@@ -389,27 +408,33 @@ code::
...
 }
 
 // or
 MyTestPoint {
     var <x, <y;
 
-    *new{ |x, y| ^super.newCopyArgs(x,y) }
+    *new { |x, y|
+		^super.newCopyArgs(x,y)
+	}
```
Comment (gusano) [inline HelpSource/Guides/WritingClasses.schelp]: @telephon and here

**Comment by telephon** [comment]
@gusano Thanks, these are fixed now.

We still have the problem #1536.

**Comment by telephon** [comment]
This turned out to have been a lot of work. There was some plain wrong information in it, too. Now is a time to give the text and structure a thorough review.

**Comment by telephon** [comment]
@gusano ok now?

File: HelpSource/Guides/WritingClasses.schelp [hunk 6]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+
+
 section:: Methods
 
-class methods are specified with the asterisk within the class method, the keyword code::this:: refers to the class. A
-class in SuperCollider is itself an object.  It is an instance of link::Classes/Class::. Instance methods are specified
-within the instance method, the keyword code::this:: refers to the instance.
+subsection:: Instance Methods
+
+Each object instance responds to its strong::instance methods::.Within instance methods, the keyword code::this:: refers to the instance itself.
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Missing space before the word `Within`.

File: HelpSource/Guides/WritingClasses.schelp [hunk 7]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+
+An instance is independent of all the other instances, unless otherwise specified.
+
 
 code::
 MyClass {
-    *classMethod { arg argument;
 
-    }
+	instanceMethod { | argument |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: There shouldn't be any white space around `argument`.

File: HelpSource/Guides/WritingClasses.schelp [hunk 8]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+		this.anotherInstanceMethod(argument)
+	}
 
-    instanceMethod { arg argument;
-
-    }
+	anotherInstanceMethod { | argument |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Same white space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 9]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+		"hello instance".postln
+	}
 }
 ::
 
+This could then be used as follows:
+
+code::
+a = MyClass.new // returns a new instance
+a.instanceMethod // posts "hello instance"
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: I would add semi-colons here, but not a big deal either way.

File: HelpSource/Guides/WritingClasses.schelp [hunk 10]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
-            ^someObject
-        },{
-            ^someOtherObject
-        })
-    }
+	someMethod {
+		^returnObject
+	}
+
+	someOtherMethod { | aBoolean |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 11]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+		if(aBoolean) {
+			^someObject
+		} {
+			^someOtherObject
+		}
+	}
+
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Remove empty line.

File: HelpSource/Guides/WritingClasses.schelp [hunk 12]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+subsection:: Class Methods
+
+An object's strong::class methods:: are defined together with the instance methods. Class methods are specified with the asterisk (teletype::*::). Within class methods, the keyword code::this:: refers to the class.
+
+A link::Classes/Class:: is itself an object. It is what all instances of it have in common.
+
+code::
+MyClass {
+	*classMethod { | argument |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 13]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+		this.anotherClassMethod(argument)
+	}
+
+	*anotherClassMethod { | argument |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 14]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+		"hello class".postln
+	}
+
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Remove empty line.

File: HelpSource/Guides/WritingClasses.schelp [hunk 15]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+	calculate { |in| ^in * 2 }
+	value { |in| ^this.calculate(in) }
+}
+
+MyClass : SomeSuperclass {
+	calculate { |in| ^in * 3 }
+}
+::
+
+If the method of the superclass is needed, it can be called by super.
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Shouldn't it be `code::super::` ?

File: HelpSource/Guides/WritingClasses.schelp [hunk 16]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+
+code::
+SomeSuperclass {
+
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Remove extra empty line.

File: HelpSource/Guides/WritingClasses.schelp [hunk 17]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+	value {
+		^100.rand
+	}
+}
+
+MyClass : SomeSuperclass {
+
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Remove extra empty line.

File: HelpSource/Guides/WritingClasses.schelp [hunk 18]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
-    }
-
-    // this is a normal constructor method
-    *new1 { arg arga,argb,argc;
-        ^super.new.init(arga,argb,argc)
-    }
-    init { arg arga,argb,argc;
-        // do initiation here
-    }
+
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Remove extra empty line.

File: HelpSource/Guides/WritingClasses.schelp [hunk 19]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+	// this is a normal constructor method
+	*new { | arga, argb, argc |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 20]
```diff
@@ -32,72 +37,146 @@ MyClass : SomeSuperclass {
...
+		^super.new.init(arga, argb, argc)
+	}
+
+	init { | arga, argb ,argc |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 21]
```diff
@@ -106,101 +185,176 @@ we are calling the code::.init:: method on that object, which is an instance met
...
-One easy way to copy the arguments passed to the instance variables when creating a class is to use newCopyArgs.  This
+One easy way to copy the arguments passed to the instance variables when creating a class is to use link::Classes/Object#*newCopyArgs::.  This
 method will copy the arguments to the instance variables in the order that the variables were defined in the class,
-starting the parent classes and working it's way down to the current class.
+starting from the parent classes and working it's way down to the current class.
 
 code::
 MyClass {
-    var <a,b,c;
+	var a,b,c;
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Add white space between a, b, c.

File: HelpSource/Guides/WritingClasses.schelp [hunk 22]
```diff
@@ -106,101 +185,176 @@ we are calling the code::.init:: method on that object, which is an instance met
...
 
-    *new { arg a,b,c;
-        ^super.newCopyArgs(a,b,c)
-    }
+	*new { | a, b, c |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 23]
```diff
@@ -106,101 +185,176 @@ we are calling the code::.init:: method on that object, which is an instance met
...
 
-MyChildClass : MyClass{
-    var <d;
+MyChildClass : MyClass {
+	var d;
 
-    *new { arg a,b,c,d;
-        ^super.newCopyArgs(a,b,c,d)
-    }
+	*new { | a, b, c, d |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 24]
```diff
@@ -106,101 +185,176 @@ we are calling the code::.init:: method on that object, which is an instance met
...
+
+Two completely unrelated objects can respond to the same messages and therefore be used together in the same code. For example, link::Classes/Function:: and link::Classes/Event:: have no common superclass apart from the general class link::Classes/Object::. But both respond to the message code::play::. Instead of inheriting all methods, you can simply implement some of the same methods in your class.
 
 code::
-Superclass {
-    calculate { arg in; in * 2 }
-    value { arg in; ^this.calculate(in) }
+MyClass {
+	var count = 0;
+	value {
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Add empty line above this one.

File: HelpSource/Guides/WritingClasses.schelp [hunk 25]
```diff
@@ -106,101 +185,176 @@ we are calling the code::.init:: method on that object, which is an instance met
...
+		^count = count + 1
+	}
 }
+// objects of this class will respond to the message "value", just like a function.
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Add empty line above this one.

File: HelpSource/Guides/WritingClasses.schelp [hunk 26]
```diff
@@ -106,101 +185,176 @@ we are calling the code::.init:: method on that object, which is an instance met
...
+a = MyClass.new;
+a.value; // returns 1
+::
+
+
+subsection::Object Composition
+Often, an object passes control to one of the objects it has in its instance variables. Because these objects can be of any kind, this is a very flexible way to achieve a wide range of functionalities. For example, a link::Classes/Button:: has an code::action:: instance variable, which may hold anything that responds to the message code::value::.
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Add empty line above this one.

File: HelpSource/Guides/WritingClasses.schelp [hunk 27]
```diff
@@ -106,101 +185,176 @@ we are calling the code::.init:: method on that object, which is an instance met
...
 
-MyClass : Superclass {
-    calculate { arg in; in * 3 }
+code::
+MyClass {
+	var action;
+	*new { |action|
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Add empty line above this one.

File: HelpSource/Guides/WritingClasses.schelp [hunk 28]
```diff
@@ -106,101 +185,176 @@ we are calling the code::.init:: method on that object, which is an instance met
...
+		^super.newCopyArgs(action)
+	}
+	value {
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Add empty line above this one.

File: HelpSource/Guides/WritingClasses.schelp [hunk 29]
```diff
@@ -235,92 +389,97 @@ methods should take only one argument to support both ways of expression consist
 
 code::
 MyClass {
-    variable_ { arg newValue;
-        variable = newValue.clip(minval,maxval);
-    }
+
+	variable_ { | newValue |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 30]
```diff
@@ -235,92 +389,97 @@ methods should take only one argument to support both ways of expression consist
...
+		variable = newValue.clip(minval,maxval);
+	}
+
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Remove empty line.

File: HelpSource/Guides/WritingClasses.schelp [hunk 31]
```diff
@@ -235,92 +389,97 @@ methods should take only one argument to support both ways of expression consist
...
+A setter method should always return the receiver. This is standard OOP practice. Never return the new value from your setter.
 
 code::
 MyClass {
-    variable_ { arg newValue;
-        myVariable = newValue;
-        ^newValue       // DO NOT DO THIS
-    }
+
+	variable_ { | newValue |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 32]
```diff
@@ -235,92 +389,97 @@ methods should take only one argument to support both ways of expression consist
...
+		myVariable = newValue;
+		^newValue       // DO NOT DO THIS
+	}
+
 }
 ::
 
-section:: External Method Files
+subsection::Constants
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Add empty line below new subsection.

File: HelpSource/Guides/WritingClasses.schelp [hunk 33]
```diff
@@ -235,92 +389,97 @@ methods should take only one argument to support both ways of expression consist
...
+
+Classes defined with code::[slot]:: can use the syntax code::myClass[...]:: which will call code::myClass.new:: and then code::this.add(each):: for each item in the square brackets.
 
 code::
-MyPointArray[slot] : RawArray {
-    center { ^Point(*this.asArray.flop.collect{ |item| item.sum / item.size } ) }
-    asArray{ ^this.collect{ |elem| elem.asArray } }
+MyClass[] {
+	var <allOfThem;
+	add { |item|
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: Add empty line above this one.

File: HelpSource/Guides/WritingClasses.schelp [hunk 34]
```diff
@@ -235,92 +389,97 @@ methods should take only one argument to support both ways of expression consist
...
 
-    *new{ |x,y| ^super.newCopyArgs(x,y) }
+	*new { |x, y|
+		^super.newCopyArgs(x, y)
+	}
 
-    printOn { arg stream;
-        stream << "MyTestPoint( " << x << ", " << y << " )";
-    }
+	printOn { | stream |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 35]
```diff
@@ -334,46 +493,52 @@ method can be used instead of code::storeOn::.
...
 
-    *new{ |x,y| ^super.newCopyArgs(x,y) }
+	*new { |x, y|
+		^super.newCopyArgs(x,y)
+	}
 
-    storeOn { arg stream;
-        stream << "MyTestPoint.new(" << x << ", " << y << ")";
-    }
+	storeOn { | stream |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 36]
```diff
@@ -334,46 +493,52 @@ method can be used instead of code::storeOn::.
...
 
-    *new{ |x,y| ^super.newCopyArgs(x,y) }
+	*new { |x, y|
+		^super.newCopyArgs(x,y)
+	}
 
-    storeArgs { arg stream;
-        ^[x,y]
-    }
+	storeArgs { | stream |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

File: HelpSource/Guides/WritingClasses.schelp [hunk 37]
```diff
@@ -334,46 +493,52 @@ method can be used instead of code::storeOn::.
...
+
+When a message is received that is undefined, the receiver calls the method code::doesNotUnderstand::. Normally this throws an error. By overriding code::doesNotUnderstand::, it is possible to catch those calls and use them (For an example, see the class definition of code::IdentityDictionary::).
 
 code::
 MyClass {
-    *new { ^super.new }
 
-    doesNotUnderstand { arg selector...args;
-        (this.class++" does not understand method "++selector);
+	doesNotUnderstand { | selector...args |
```
Comment (patrickdupuis) [inline HelpSource/Guides/WritingClasses.schelp]: White space issue here.

**Comment by patrickdupuis** [comment]
We should try and cherry-pick this into 3.9 before the next release if possible.

**Comment by telephon** [comment]
see #3605

**Comment by patrickdupuis** [comment]
I don't understand why you closed this PR and opened #3605. You could have made those changes here, merged this PR into `develop`, and them cherry-picked it into `3.9`. You should merge the changes you made in #3605 here, and continue.

To compare : https://github.com/telephon/supercollider/compare/topic-help-improve-writing-classed...telephon:topic-fix-writing-classes

**Pull request closed without merging**